mantidproject/mantid
# Add a new `ComplexImage` property to MaxEnt

Added a new input property, `ComplexImage`, so that scientists can force the image to be real.

**To test:**

Run the following script:

```
from math import pi, sin, cos
X = []
YRe = []
YIm = []
E = []
N = 500
w = 3
for i in range(0,N):
    x = 2*pi*i/N
    X.append(x)
    YRe.append(cos(w*x))
    YIm.append(sin(w*x))
    E.append(0.1)
CreateWorkspace(OutputWorkspace='ws',DataX=X+X,DataY=YRe+YIm,DataE=E+E,NSpec=2)
evolChi1, evolAngle1, image1, data1 = MaxEnt(InputWorkspace='ws', ComplexData=True, ComplexImage=True, chiTarget=2*N, A=0.001)
evolChi2, evolAngle2, image2, data2 = MaxEnt(InputWorkspace='ws', ComplexData=True, ComplexImage=False, chiTarget=2*N, A=0.001)
```

which creates a workspace with a function `cos(w0 * t) + i sin(w0 * t)`, for which the Fourier transform is a delta peak `delta (w-w0)`. Both executions should give similar results. Note that if you change the input signal to `sin(w0 * t) + i cos(w0 * t)`, like this:

```
from math import pi, sin, cos
X = []
YRe = []
YIm = []
E = []
N = 500
w = 3
for i in range(0,N):
    x = 2*pi*i/N
    X.append(x)
    YIm.append(cos(w*x))
    YRe.append(sin(w*x))
    E.append(0.1)
CreateWorkspace(OutputWorkspace='ws',DataX=X+X,DataY=YRe+YIm,DataE=E+E,NSpec=2)
evolChi1, evolAngle1, image1, data1 = MaxEnt(InputWorkspace='ws', ComplexData=True, ComplexImage=True, chiTarget=2*N, A=0.001)
evolChi2, evolAngle2, image2, data2 = MaxEnt(InputWorkspace='ws', ComplexData=True, ComplexImage=False, chiTarget=2*N, A=0.001)
```

the algorithm will fail to converge, as it will be trying to reconstruct a complex image using its real part only.

Fixes #15553



---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
